src/irmin-pack/unix/ext.ml

```diff
@@ -218,44 +211,242 @@ module Maker (Config : Conf.S) = struct
+                Error (`Corrupted_gc_result_file (Fmt.str "Exited %d %s" n str))
+            | Lwt_unix.WSIGNALED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Signaled %d" n))
+            | Lwt_unix.WSTOPPED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d" n))
+            | Lwt_unix.WSIGNALED n, Error (`Gc_process_error str) ->
+                Error (`Gc_process_error (Fmt.str "Signaled %d %s" n str))
+            | Lwt_unix.WSTOPPED n, Error (`Gc_process_error str) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d %s" n str))
+            | Lwt_unix.WEXITED _, Ok _
+            | Lwt_unix.WSIGNALED _, Ok _
+            | Lwt_unix.WSTOPPED _, Ok _ ->
+                assert false
+
+          let finalise ~wait t =
+            let { next_generation; pid; unlink } = Option.get t.during_gc in
+            let wait_flags = if wait then [] else [ Unix.WNOHANG ] in
+            let root = Conf.root t.config in
+            let go status =
+              let gc_output =
+                File_manager.read_gc_output ~unlink:false ~root
+                  ~generation:next_generation
+              in
+              let result =
+                match (status, gc_output) with
+                | Lwt_unix.WEXITED 0, Ok last_copied_offset ->
+                    Result.bind
+                      (copy_latest_newies ~generation:next_generation
+                         ~last_copied_offset ~root t) (fun () ->
+                        swap_and_purge ~unlink ~generation:next_generation
+                          ~last_copied_offset t)
+                | _ -> gc_errors (status, gc_output)
+              in
+              t.during_gc <- None;
+              Exit.remove pid;
+              result
+            in
+            if t.during_batch then Lwt.return_error `Gc_forbidden_during_batch
+            else
+              let* pid, status = Lwt_unix.waitpid wait_flags pid in
+              (* Do not block if no child has died yet. In this case the waitpid
+                 returns immediately with a pid equal to 0. *)
+              if (not wait) && pid = 0 then Lwt.return_ok false
+              else
+                (* If wait or the child died, then finalise the gc. *)
+                go status |> Result.map (fun () -> true) |> Lwt.return
+
+          let start_and_maybe_finalise ~unlink ~auto_finalise t commit_key =
+            let open Lwt_result.Syntax in
+            match start ~unlink t commit_key with
+            | Error e -> Lwt.return_error e
+            | Ok () ->
+                if auto_finalise then
+                  let* (_ : bool) = finalise ~wait:false t in
+                  Lwt.return_ok ()
+                else Lwt.return_ok ()
```

Comment:
I think you can remove this function can directly call `start`. It's not possible for `start` to finish right away